### DISCUSSION TITLE
Biscuit v3.0.0 - New look & features

### DISCUSSION THREAD

**Discussion by tomlin7** [tomlin7/biscuit#465] [Announcements]
![image](https://github.com/user-attachments/assets/51e608e3-c8f1-4d2d-8ecf-fb51803d037b)

This latest `v3.0` major release includes many changes, notably a complete redesign of the editor's appearance. Yet, there are many important PRs and feature requests that couldn't make it to this release!

Many contributors have added new features to the project during the hacktoberfest season: [see the list](https://github.com/tomlin7/biscuit/issues?q=label%3Ahacktoberfest-accepted+is%3Aclosed)

![image](https://github.com/user-attachments/assets/49258dec-5b15-462a-a056-5401ac0b73d9)

## Quick Install:
```
pip install --upgrade biscuit-editor
```
then just run `biscuit path/to/directory` to launch!

For a detailed list of features of Biscuit, go through the [feature section](https://github.com/tomlin7/biscuit?tab=readme-ov-file#features).

## What's Changed
* Layout redesign, new iconset, new themes by @tomlin7 in https://github.com/tomlin7/biscuit/pull/391
* Extension Search bar implementation by @Kabiirk in https://github.com/tomlin7/biscuit/pull/439
* Extension details pane by @tomlin7 in https://github.com/tomlin7/biscuit/pull/442
* Added [editorconfig](https://editorconfig.org/) support  by @tomlin7 in https://github.com/tomlin7/biscuit/pull/433
* Improved welcome page by @tomlin7 in https://github.com/tomlin7/biscuit/pull/394
* Text editor Context Menu by @tomlin7 in https://github.com/tomlin7/biscuit/pull/399
* Restoring most recently closed tabs (#400) by @GooglyBlox in https://github.com/tomlin7/biscuit/pull/401
* New command palette design by @tomlin7 in https://github.com/tomlin7/biscuit/pull/446
* Relative line numbering by @nkapila6 in https://github.com/tomlin7/biscuit/pull/420
* Installed extensions filter for the extension center by @saai-syvendra in https://github.com/tomlin7/biscuit/pull/423
* Floating opened editors list  by @tomlin7 in https://github.com/tomlin7/biscuit/pull/395
* Submenus support for biscuit menus by @tomlin7 in https://github.com/tomlin7/biscuit/pull/397
* Make command palette scrollable, other UI improvements by @tomlin7 in https://github.com/tomlin7/biscuit/pull/447
* Added CLI path argument (`biscuit <dir>`) by @moonlander101 in https://github.com/tomlin7/biscuit/pull/404
* More keybindings by @tomlin7 in https://github.com/tomlin7/biscuit/pull/449
* Recent folders/files: Bump newly opened to top if exists in db by @tomlin7 in https://github.com/tomlin7/biscuit/pull/454
* Biscuit Workspaces by @tomlin7 in https://github.com/tomlin7/biscuit/pull/426
* Extensions user-level storage by @tomlin7 in https://github.com/tomlin7/biscuit/pull/430
* Insert final new line on save if `insert_final_newline` is true by @tomlin7 in https://github.com/tomlin7/biscuit/pull/432

### Fixes
* docs: Getting started by @tomlin7 in https://github.com/tomlin7/biscuit/pull/419
* docs: Added Arch Linux commands to contributing guidelines and various minor changes by @shirsakm in https://github.com/tomlin7/biscuit/pull/437
* fix: Update Breadcrumbs on editor switch/editor close by @tomlin7 in https://github.com/tomlin7/biscuit/pull/444
* fix: Fixed app crashes due to hover renderer shrink property by @tomlin7 in https://github.com/tomlin7/biscuit/pull/411

## New Contributors
* @GooglyBlox made their first contribution in https://github.com/tomlin7/biscuit/pull/401
* @moonlander101 made their first contribution in https://github.com/tomlin7/biscuit/pull/404
* @nkapila6 made their first contribution in https://github.com/tomlin7/biscuit/pull/420
* @saai-syvendra made their first contribution in https://github.com/tomlin7/biscuit/pull/423
* @shirsakm made their first contribution in https://github.com/tomlin7/biscuit/pull/437
* @Kabiirk made their first contribution in https://github.com/tomlin7/biscuit/pull/439

**Full Changelog**: https://github.com/tomlin7/biscuit/compare/v2.99.22...v3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/tomlin7/biscuit/releases/tag/v3.0.0'>Biscuit v3.0.0 - New look & features</a>.</em>

**Comment by giohappy** [comment]
@tomlin7 you're the author of a Python DAP client library. Do you have plans to integrate it in Biscuit?

I'm asking not only for the features DAP provides to debug multiple languages, but also having the option to attach to a process, which apparently isn't implemented in Biscuit yet.

**Reply by tomlin7** [reply]
@giohappy You might've noticed that Biscuit only has a python debugger implemented right now. Although the debugger part of the editor is extensible, you can publish extensions for different debuggers: [Built in python bdb implementation](https://github.com/tomlin7/biscuit/blob/main/src%2Fbiscuit%2Fdebugger%2Fpython.py).

Now, this *can* potentially be an OK solution for the problem. But it requires me/others to contribute extensions for each of the debuggers out there, each of which will implement the requirements for Biscuit debugger base manually.

DAP kind of solves this easily because debug adapters are available, and we can do a unified implementation. That DAP client you mentioned was made for Biscuit. But improving the coding agent part kind of kept me busy, so never got to finally bring it in.

**Reply by giohappy** [reply]
So an extension could be made to implement the DAP debugger. Nice. 

My question was if you planned to make one for your own dap client library in the future.

**Reply by tomlin7** [reply]
@giohappy Yes, I mentioned that as well. I'm planning to do it soon. The DAP client was made for integrating into biscuit. 

Once this is done, any up-to-date debug adapter should be working with the IDE.

About the attaching to process part: it's not done yet, but is planned.

**Reply by tomlin7** [reply]
@giohappy  I have completely rewritten the dap client library to be Sans-IO.

You can have a look at the sandbox implementation as a reference for easily integrating the client into GUI applications.

https://github.com/tomlin7/debug-adapter-client
